// dim-console/dist/dim-console.js
// Generated by LiveScript 1.5.0
var dim, dimConsole;
dim = require('chalk').dim;
dimConsole = {
  stdoutOutput: '',
  stderrOutput: '',
  output: '',
  console: {
    log: function(text){
      var textWithNewline;
      textWithNewline = text + "\n";
      dimConsole.stdoutOutput += textWithNewline;
      dimConsole.output += textWithNewline;
      return console.log(dim(text.trimRight()));
    },
    error: function(text){
      var textWithNewline;
      textWithNewline = text + "\n";
      dimConsole.stderrOutput += textWithNewline;
      dimConsole.output += textWithNewline;
      return console.log(dim(text.trimRight()));
    }
  },
  process: {
    stdout: {
      write: function(text){
        dimConsole.stdoutOutput += text;
        dimConsole.output += text;
        return process.stdout.write(dim(text));
      }
    },
    stderr: {
      write: function(text){
        dimConsole.stderrOutput += text;
        dimConsole.output += text;
        return process.stderr.write(dim(text));
      }
    }
  },
  reset: function(){
    dimConsole.output = '';
    dimConsole.stdoutOutput = '';
    return dimConsole.stderrOutput = '';
  }
};
module.exports = dimConsole;
