// exo-sync/dist/service-syncer.js
// Generated by LiveScript 1.5.0
var EventEmitter, fs, yaml, ObservableProcess, path, ServiceSyncer;
EventEmitter = require('events').EventEmitter;
fs = require('fs');
yaml = require('js-yaml');
ObservableProcess = require('observable-process');
path = require('path');
ServiceSyncer = (function(superclass){
  var prototype = extend$((import$(ServiceSyncer, superclass).displayName = 'ServiceSyncer', ServiceSyncer), superclass).prototype, constructor = ServiceSyncer;
  function ServiceSyncer(name, config){
    this.name = name;
    this.config = config;
    this.write = bind$(this, 'write', prototype);
    this.start = bind$(this, 'start', prototype);
  }
  ServiceSyncer.prototype.start = function(done){
    var x$, this$ = this;
    switch (false) {
    case !!this._isExternalService():
      return done(null, 0);
    }
    x$ = new ObservableProcess("git pull", {
      cwd: this.config.root,
      stdout: {
        write: this.write
      },
      stderr: {
        write: this.write
      }
    });
    x$.on('ended', function(exitCode){
      switch (false) {
      case exitCode !== 0:
        this$.write('Sync ok');
        break;
      default:
        this$.write('Sync error');
      }
      return done(null, exitCode);
    });
    return x$;
  };
  ServiceSyncer.prototype._isExternalService = function(){
    return this.config.root.split('/')[0] === '..';
  };
  ServiceSyncer.prototype.write = function(text){
    return this.emit('output', {
      name: this.name,
      text: text
    });
  };
  return ServiceSyncer;
}(EventEmitter));
module.exports = ServiceSyncer;
function bind$(obj, key, target){
  return function(){ return (target || obj)[key].apply(obj, arguments) };
}
function extend$(sub, sup){
  function fun(){} fun.prototype = (sub.superclass = sup).prototype;
  (sub.prototype = new fun).constructor = sub;
  if (typeof sup.extended == 'function') sup.extended(sub);
  return sub;
}
function import$(obj, src){
  var own = {}.hasOwnProperty;
  for (var key in src) if (own.call(src, key)) obj[key] = src[key];
  return obj;
}
