// exprestive/dist/url_formatter.js
// Generated by CoffeeScript 1.10.0
(function() {
  var ReverseRoute, URLFormatter,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    slice = [].slice;

  ReverseRoute = require('./reverse_route');

  URLFormatter = (function() {
    function URLFormatter(template) {
      this.template = template;
      this.getRoute = bind(this.getRoute, this);
    }

    URLFormatter.prototype.replaceParams = function() {
      var i, len, name, param, params, ref, url, value;
      params = 1 <= arguments.length ? slice.call(arguments, 0) : [];
      url = this.template;
      if (typeof params[0] === 'object') {
        ref = params[0];
        for (name in ref) {
          value = ref[name];
          url = url.replace(":" + name, value);
        }
      } else {
        for (i = 0, len = params.length; i < len; i++) {
          param = params[i];
          url = url.replace(/:[^\/]+/, param);
        }
      }
      return url;
    };

    URLFormatter.prototype.getRoute = function(httpMethod, params) {
      return new ReverseRoute({
        path: this.replaceParams.apply(this, params),
        method: httpMethod
      });
    };

    return URLFormatter;

  })();

  module.exports = URLFormatter;

}).call(this);
